lib/v4/MessageBuilder.test.ts

```diff
@@ -0,0 +1,310 @@
+      '.': zod.string(),
+      './*': zod.string(),
+    });
+
+    const messageBuilder = createMessageBuilder({ error: errorMap });
+
+    try {
+      schema.parse(
+        {
+          '.': 123,
+          './*': false,
+        },
+        {
+          reportInput: true,
+        }
+      );
+    } catch (err) {
+      if (isZodErrorLike(err) && isNonEmptyArray(err.issues)) {
+        const message = messageBuilder(err.issues);
+        expect(message).toMatchInlineSnapshot(
+          `"Validation error: Expected string, received number at "."; Expected string, received boolean at "./*""`
+        );
+      }
+    }
+  });
+
+  test('handles zod.function() argument errors', () => {
+    const fn = zod
+      .function({
+        input: [zod.number()],
+      })
+      .implement((num) => num * 2);
+
+    const messageBuilder = createMessageBuilder({ error: errorMap });
+
+    try {
+      // @ts-expect-error Intentionally wrong to exercise runtime checking
+      fn('foo');
+    } catch (err) {
+      if (isZodErrorLike(err) && isNonEmptyArray(err.issues)) {
+        const message = messageBuilder(err.issues);
+        expect(message).toMatchInlineSnapshot(
+          `"Validation error: Expected number at index 0"`
+        );
+      }
+    }
+  });
+
+  test('handles zod.function() return value errors', () => {
+    const fn = zod
+      .function({
+        output: zod.number(),
+      })
+      // @ts-expect-error Intentionally wrong to exercise runtime checking
+      .implement(() => 'foo');
+
+    const messageBuilder = createMessageBuilder({ error: errorMap });
+
+    try {
+      fn();
+    } catch (err) {
+      if (isZodErrorLike(err) && isNonEmptyArray(err.issues)) {
+        const message = messageBuilder(err.issues);
+        expect(message).toMatchInlineSnapshot(
+          `"Validation error: Expected number"`
+        );
+      }
+    }
+  });
+
+  test('respects `includePath` prop when set to `false`', () => {
+    const schema = zod.object({
+      name: zod.string().min(3),
+    });
+
+    const messageBuilder = createMessageBuilder({
+      error: createErrorMap({
+        includePath: false,
+      }),
+    });
+
+    try {
+      schema.parse({ name: 'jo' });
+    } catch (err) {
+      if (isZodErrorLike(err) && isNonEmptyArray(err.issues)) {
+        const message = messageBuilder(err.issues);
+        expect(message).toMatchInlineSnapshot(
+          `"Validation error: String must contain at least 3 character(s)"`
+        );
+      }
+    }
+  });
+
+  test("returns zod's native error messages by default", () => {
+    const schema = zod.object({
+      name: zod.string().min(3),
+    });
+
+    const messageBuilder = createMessageBuilder({
+      // note that we are not passing `error` here
+      // so it will use zod's default error map
+    });
+
+    try {
+      schema.parse({ name: 'jo' });
+    } catch (err) {
+      if (isZodErrorLike(err) && isNonEmptyArray(err.issues)) {
```

Comment:
question: should we always go straight to `expect` statement? I'm saying because in case the "if" statement for some weird reason is not `true`, we might not reach the `expect` statement, therefore the test will be passing, but simply because no expectation fails.

Reply:
Hm, we can definitely improve the test code. This is the same as before though (i.e. before this PR), so I wonder whether in scope.